### ISSUE TITLE
Don't get it working in CMS8

### ISSUE THREAD

**Issue by haassie** [FriendsOfTYPO3/openid#18]
I tried the plugin with 2 different openid providers, but both are not working the the CMS8 version. On CMS7 it is working fine (older version of plugin). 

Does anyone get the plugin working in CMS8? As far as I have debugged now, I get an error 
`No matching endpoint found after discovering https://mydomain.com/openidserver/users/username: Required type http://openid.net/signon/1.0 not present`

**Comment by liayn** [comment]
Seems like there is a fallback to version 1.0 of the protocol which does not work with your providers as those most likely use 2.0. I just wonder why the lib fails to connect with 2.0

I'm waiting to receive another test-account, so I can take a closer look myself.

**Comment by haassie** [comment]
For your information: I tried with our Atlassian Crowd and with Yahoo OpenID. Both has the same problem.

**Comment by haassie** [comment]
Some extra information: Version 7.6.1 (with some changes in the database handling of the contraints in the getUserRecord method) was working correctly in the CMS8 installation.

**Comment by liayn** [comment]
Thanks for the update. So lets see what's different in those versions.

**Comment by oaltmann** [comment]
@haassie: Can you tell me your changes in the database handling?

**Comment by haassie** [comment]
@oaltmann the main problem was the missing AND in the `getUserRecord` method. But if you cherrypick https://github.com/FriendsOfTYPO3/openid/commit/94bac9a24a6af4c103dfa0837ef2a75d6def3558 I think it should work.

**Comment by oaltmann** [comment]
Thanks for your info. 
I debugged the same function and i found my problem. The `$record` in the `getUserRecord` function in the OpenIdService.php is null. Because the `$this->authenticationInformation['db_user']['check_pid_clause']` and `$this->authenticationInformation['db_user']['enable_clause']` return a object not a string. I don't no why. I use openId 7.6.4. Do you have a solution for this?

**Comment by liayn** [comment]
@oaltmann That's an extremely helpful finding thanks. This is related to the change to doctrine with CMS 8.
The authInfo clauses are now doctrine objects and not plain strings. We need to adjust that.

**Comment by liayn** [comment]
@oaltmann  I pushed a fix, please test if it works

**Comment by oaltmann** [comment]
@liayn: Your fix with version 8.0.0 does not work. But i put your fix in Version 7.6.4. It works very well in TYPO3 8.7.0. Do you know the different between this versions?

**Comment by liayn** [comment]
??? I don't get what you are saying. I talk about this commit https://github.com/FriendsOfTYPO3/openid/commit/52dbdf003123f404f2d8ac8fce91f2abdcfd6e98 which will only work in CMS 8, since $queryBuilder and stuff are only available there.

**Comment by oaltmann** [comment]
Ok. Sorry.
If i use the master with your fix (Version 8.0.0), it does not work in TYPO3 8.7.0. 
If i use the Version 7.6.4 (tag) and i put your fix in this Version it works very well.

**Comment by liayn** [comment]
@oaltmann I will test this eventually on my own now. I still don't understand how the last commit shall apply on the 7.6.4 version, since this is totally different code there. How do you do it? Cherry-pick?

**Comment by oaltmann** [comment]
I don't pick the 7.6.4 Version from github. I use the version from a own current TYPO3 7.6 Projekt and change only the OpenidService.php and then it works. (The 7.6.4 Version is from the TYPO3 Repository).
The Cherry-pick  `52dbdf0` Version does not work in TYPO3 8.7.0.

**Comment by liayn** [comment]
And what do you change in there exactly?

**Comment by oaltmann** [comment]
I change the `getUserRecord` to this:
```
    /**
     * Gets user record for the user with the OpenID provided by the user
     *
     * @param string $openIDIdentifier OpenID identifier to search for
     * @return array Database fields from the table that corresponds to the current login mode (FE/BE)
     */
    protected function getUserRecord($openIDIdentifier)
    {
        $record = null;
        try {
            $openIDIdentifier = $this->normalizeOpenID($openIDIdentifier);
            // $openIDIdentifier always has a trailing slash
            // but tx_openid_openid field possibly not so check for both alternatives in database
            $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable($this->authenticationInformation['db_user']['table']);
            $queryBuilder->getRestrictions()->removeAll();
            $record = $queryBuilder
                ->select('*')
                ->from($this->authenticationInformation['db_user']['table'])
                ->where(
                    $queryBuilder->expr()->in(
                        'tx_openid_openid',
                        $queryBuilder->createNamedParameter(
                            [$openIDIdentifier, rtrim($openIDIdentifier, '/')],
                            Connection::PARAM_STR_ARRAY
                        )
                    ),
                    $this->authenticationInformation['db_user']['check_pid_clause'],
                    $this->authenticationInformation['db_user']['enable_clause']
                )
                ->execute()
                ->fetch();

            if ($record) {
                // Make sure to work only with normalized OpenID during the whole process
                $record['tx_openid_openid'] = $this->normalizeOpenID($record['tx_openid_openid']);
            }
        } catch (Exception $e) {
            // This should never happen and generally means hack attempt.
            // We just log it and do not return any records.
            $this->writeLog($e->getMessage());
        }

        // Hook to modify the user record, e.g. to register a new user
        if (isset($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['openid']['getUserRecord']) && is_array($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['openid']['getUserRecord'])) {
            $_params = array(
                'record' => &$record,
                'response' => $this->openIDResponse,
                'authInfo' => $this->authenticationInformation
            );
            foreach ($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['openid']['getUserRecord'] as $funcName) {
                GeneralUtility::callUserFunction($funcName, $_params, $this);
            }
        }

        return $record;
    }
```

And i change the namespaces to this:
```
use TYPO3\CMS\Core\Crypto\Random;
use TYPO3\CMS\Core\Database\Connection;
use TYPO3\CMS\Core\Database\ConnectionPool;
use TYPO3\CMS\Core\Service\AbstractService;
use TYPO3\CMS\Core\TimeTracker\TimeTracker;
use TYPO3\CMS\Core\Utility\ExtensionManagementUtility;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Core\Authentication\AbstractUserAuthentication;
use TYPO3\CMS\Core\Utility\HttpUtility;
```

**Comment by liayn** [comment]
Okay thanks. So that's the proof that our DB logic is obviously correct. I suspect that changes to the openid lib are responsible then, which went into the v8 version of ext:openid as well.

**Comment by liayn** [comment]
@oaltmann I did a lot of debugging now.
I tried with my identity `https://markus.klein.id.tankom.de/` and the whole thing fails with

```
PHP Warning: fsockopen(): Peer certificate CN=`*.id.tankom.de' did not match expected CN=`markus.klein.id.tankom.de' in /mnt/hgfs/typo3ext/openid/lib/php-openid/Auth/Yadis/PlainHTTPFetcher.php line 106
```

So clearly the certificate is the root cause here, as it is not accepted by openssl.
Also surfing to the url in the browser gives the same result.

I'm totally unsure though if `*.id.tankom.de` should match `markus.klein.id.tankom.de` or not. I've to research on that.

**Comment by oaltmann** [comment]
@liayn I send you a email. Because your identity is wrong. I create a new account for you. Sorry for that.

**Comment by liayn** [comment]
Okay after searching for hours I pushed a fix for a hidden breaking change in the core, which caused the wizard to not work properly.
I still get an error now, but I got one step further. Halleluja.

**Comment by liayn** [comment]
Finally got it working. Released v8.0.1 now.
